subscriptions/views.py

```diff
@@ -0,0 +1,223 @@
+from __future__ import division
+
+from datetime import datetime
+import json
+import smtplib
+
+from django.conf import settings
+from django.contrib import messages
+from django.core import mail
+from django.core.urlresolvers import reverse_lazy
+from django.shortcuts import redirect
+from django.template.loader import render_to_string
+from django.views.generic import DetailView, FormView, DeleteView, View
+from django.views.decorators.csrf import csrf_exempt
+from django.views.decorators.http import require_POST
+from django.http import HttpResponseRedirect, HttpResponse
+import stripe
+
+from mapit_mysociety_org.mixins import NeverCacheMixin
+from .forms import SubsForm
+from .models import Subscription
+
+
+class StripeObjectMixin(object):
+    def get_object(self):
+        try:
+            sub = self.subscription = Subscription.objects.get(user=self.request.user)
+            return stripe.Subscription.retrieve(sub.stripe_id, expand=['customer.default_source'])
+        except Subscription.DoesNotExist:
+            # There will be existing accounts with no subscription object
+            return None
+        except stripe.error.InvalidRequestError:
+            # If the subscription is missing at the Stripe end, assume it's
+            # gone and remove here too
+            sub.delete()
+            return None
+
+
+class SubscriptionView(StripeObjectMixin, NeverCacheMixin, DetailView):
+    model = Subscription
+    context_object_name = 'stripe'
+
+    def get_context_data(self, **kwargs):
+        kwargs['STRIPE_PUBLIC_KEY'] = settings.STRIPE_PUBLIC_KEY
+        context = super(SubscriptionView, self).get_context_data(**kwargs)
+        if not self.object:
+            return context
+
+        context['quota_status'] = self.subscription.redis_status()
+
+        data = self.object
+        for fld in ['current_period_start', 'current_period_end', 'created', 'start']:
+            data[fld] = datetime.fromtimestamp(data[fld])
+
+        # Amounts in pounds, not pence
+        data['plan']['amount'] /= 100
+
+        # Calculate actual amount paid, including discount
+        if data['discount'] and data['discount']['coupon'] and data['discount']['coupon']['percent_off']:
+            context['actual_paid'] = data['plan']['amount'] * (100 - data['discount']['coupon']['percent_off']) / 100
+        else:
+            context['actual_paid'] = data['plan']['amount']
+
+        return context
+
+
+class SubscriptionUpdateMixin(object):
+    def update_subscription(self, form):
```

Comment:
I think there's a bug with the charitable flag not being cleared somewhere here.

 1. Sign up as a new user, select free tier by ticking charitable box and individual radio button
 1. Go to change plan page, untick the charitable box and enter payment information
 1. Money taken, redirected to success page, but charitable status and "It costs you £0/mth." remains.